extensions/amp-autocomplete/validator-amp-autocomplete.protoascii

```diff
@@ -0,0 +1,102 @@
+#
+# Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the license.
+#
+
+tags: {  # amp-autocomplete
+  html_format: AMP
+  tag_name: "SCRIPT"
+  extension_spec: {
+    name: "amp-autocomplete"
+    version: "0.1"
+    version: "latest"
+  }
+  attr_lists: "common-extension-attrs"
+}
+tags: {  # <amp-autocomplete>
+  html_format: AMP
+  tag_name: "AMP-AUTOCOMPLETE"
+  requires_extension: "amp-autocomplete"
+  requires_extension: "amp-form"
+  attrs: {
+    name: "filter"
+    value_casei: "substring"
+    value_casei: "token-prefix"
+    value_casei: "prefix"
+    value_casei: "fuzzy"
+    value_casei: "custom"
+    value_casei: "none"
+    trigger: {
+      if_value_regex: "custom"
+      also_requires_attr: "filter-expr"
+    }
+  }
+  attrs: {
+    name: "filter-expr"
+    requires_extension: "amp-bind"
+  }
+  attrs: { name: "min-characters" }
+  attrs: { name: "max-entries" }
+  attrs: { name: "suggest-first" }
+  attrs: { name: "submit-on-enter" }
+  attr_lists: "extended-amp-global"
+  spec_url: "https://www.ampproject.org/docs/reference/components/amp-autocomplete"
+  amp_layout: {
+    supported_layouts: CONTAINER
+  }
+  mandatory_ancestor: "FORM"
+  reference_points: {
+    tag_spec_name: "amp-autocomplete > input"
+    mandatory: true
+    unique: true
+  }
+  reference_points: {
+    tag_spec_name: "amp-autocomplete JSON"
+    unique: true
+  }
```

Comment:
This will not work, unfortunately. The idea of reference points is to allow additional attributes on child tags, not define a set of allowed child tags. For instance, if you want to require a `default` attribute on one child.

I suspect what you are trying to accomplish is to say:

- Exactly one of my children must be an input tag with some requirements.
- Exactly one of my children must be a `script application/json`.
- No other children are allowed.

I don't think there is a mechanism currently built in validator configs that is designed for exactly this purpose, but this might work. In the `amp-autocomplete` tagspec, define the allowed child tag names like so:

```
child_tags: {
  mandatory_num_child_tags = 2;
  child_tag_name_oneof: 'script'
  child_tag_name_oneof: 'input'
}
```

Then, for the `input` tag, modify the other `input` tagspecs to include `disallowed_ancestor: 'amp-autocomplete'` so only your own input tagspec will match.

I think this will get almost all of the requirements you are looking for. The only thing it allows that you don't want is that the user provides two input tags or two script tags, rather than one of each, however I think it's probably OK if the validator doesn't make this guarantee.

What do you think?

Reply:
Thanks for your advice on this. You're right that I want to say that exactly one of my children must be an `input` tag with some requirements; however, `script` is optional. Is there a way to say that you can include zero or one `script application/json`?

Reply:
Ah, I see if we follow the validator convention in amp-next-page the two types of uses for autocomplete (with and without script as a child) should be two distinct tags: https://github.com/ampproject/amphtml/blob/master/extensions/amp-next-page/validator-amp-next-page.protoascii

Is that what you would recommend?

Reply:
Got it. I don't believe the rules currently support this. If you feel strongly that we are strict here, we can talk about how to extend the rules.

Alternatively, you could just remove the `mandatory_num_child_tags` line and the validator will only ensure the script/input children, but not the number.